File: 2024/day08/node/solution.js
import { readFileSync } from 'fs';

function parseInput(filename) {
    const grid = readFileSync(filename, 'utf-8').trim().split('\n');
    const rows = grid.length;
    const cols = grid[0]?.length ?? 0;

    // Group antenna positions by frequency
    const antennas = new Map();
    grid.forEach((row, r) => {
        [...row].forEach((ch, c) => {
            if (ch !== '.') {
                if (!antennas.has(ch)) {
                    antennas.set(ch, []);
                }
                antennas.get(ch).push([r, c]);
            }
        });
    });

    return { rows, cols, antennas };
}

function part1() {
    const { rows, cols, antennas } = parseInput('../input.txt');
    const antinodes = new Set();

    for (const positions of antennas.values()) {
        // For each pair of antennas with same frequency
        for (let i = 0; i < positions.length; i++) {
            for (let j = i + 1; j < positions.length; j++) {
                const [r1, c1] = positions[i];
                const [r2, c2] = positions[j];

                // Calculate the two antinodes (2:1 distance ratio)
                // Antinode beyond antenna 1 (away from antenna 2)
                const ar1 = 2 * r1 - r2;
                const ac1 = 2 * c1 - c2;
                // Antinode beyond antenna 2 (away from antenna 1)
                const ar2 = 2 * r2 - r1;
                const ac2 = 2 * c2 - c1;

                // Add if within bounds
                if (ar1 >= 0 && ar1 < rows && ac1 >= 0 && ac1 < cols) {
                    antinodes.add(`${ar1},${ac1}`);
                }
                if (ar2 >= 0 && ar2 < rows && ac2 >= 0 && ac2 < cols) {
                    antinodes.add(`${ar2},${ac2}`);
                }
            }
        }
    }

    return antinodes.size;
}

function part2() {
    const { rows, cols, antennas } = parseInput('../input.txt');
    const antinodes = new Set();

    for (const positions of antennas.values()) {
        // For each pair of antennas with same frequency
        for (let i = 0; i < positions.length; i++) {
            for (let j = i + 1; j < positions.length; j++) {
                const [r1, c1] = positions[i];
                const [r2, c2] = positions[j];
                const dr = r2 - r1;
                const dc = c2 - c1;

                // Extend in both directions along the line
                // Direction 1: from antenna 1 towards and beyond antenna 2
                let r = r1, c = c1;
                while (r >= 0 && r < rows && c >= 0 && c < cols) {
                    antinodes.add(`${r},${c}`);
                    r += dr;
                    c += dc;
                }

                // Direction 2: from antenna 1 away from antenna 2
                r = r1 - dr;
                c = c1 - dc;
                while (r >= 0 && r < rows && c >= 0 && c < cols) {
                    antinodes.add(`${r},${c}`);
                    r -= dr;
                    c -= dc;
                }
            }
        }
    }

    return antinodes.size;
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());

File: 2024/day08/python/solution.py
#!/usr/bin/env python3
from collections import defaultdict
from itertools import combinations

def parse_input(filename: str) -> tuple[int, int, dict[str, list[tuple[int, int]]]]:
    """Parse grid and group antenna positions by frequency."""
    with open(filename) as f:
        grid = [line.rstrip('\n') for line in f]

    rows = len(grid)
    cols = len(grid[0]) if grid else 0

    # Group antenna positions by frequency
    antennas = defaultdict(list)
    for r, row in enumerate(grid):
        for c, ch in enumerate(row):
            if ch != '.':
                antennas[ch].append((r, c))

    return rows, cols, dict(antennas)

def part1() -> int:
    """Count antinodes at 2:1 distance ratio from antenna pairs."""
    rows, cols, antennas = parse_input('../input.txt')
    antinodes = set()

    for positions in antennas.values():
        # For each pair of antennas with same frequency
        for (r1, c1), (r2, c2) in combinations(positions, 2):
            # Calculate the two antinodes (2:1 distance ratio)
            # Antinode beyond antenna 1 (away from antenna 2)
            ar1, ac1 = 2 * r1 - r2, 2 * c1 - c2
            # Antinode beyond antenna 2 (away from antenna 1)
            ar2, ac2 = 2 * r2 - r1, 2 * c2 - c1

            # Add if within bounds
            if 0 <= ar1 < rows and 0 <= ac1 < cols:
                antinodes.add((ar1, ac1))
            if 0 <= ar2 < rows and 0 <= ac2 < cols:
                antinodes.add((ar2, ac2))

    return len(antinodes)

def part2() -> int:
    """Count all grid positions collinear with antenna pairs."""
    rows, cols, antennas = parse_input('../input.txt')
    antinodes = set()

    for positions in antennas.values():
        # For each pair of antennas with same frequency
        for (r1, c1), (r2, c2) in combinations(positions, 2):
            dr, dc = r2 - r1, c2 - c1

            # Extend in both directions along the line
            # Direction 1: from antenna 1 towards and beyond antenna 2
            r, c = r1, c1
            while 0 <= r < rows and 0 <= c < cols:
                antinodes.add((r, c))
                r += dr
                c += dc

            # Direction 2: from antenna 1 away from antenna 2
            r, c = r1 - dr, c1 - dc
            while 0 <= r < rows and 0 <= c < cols:
                antinodes.add((r, c))
                r -= dr
                c -= dc

    return len(antinodes)

if __name__ == '__main__':
    print('Part 1:', part1())
    print('Part 2:', part2())
